Quadriga/src/main/java/edu/asu/spring/quadriga/service/impl/passthroughproject/PassThroughProjectManager.java

```diff
@@ -0,0 +1,141 @@
+    private IWorkspaceDAO workspaceDao;
+
+    @Autowired
+    private INetworkManager networkManager;
+
+    @Autowired
+    private IExternalWSManager externalWSManager;
+
+    @Autowired
+    private IPassThroughProjectDAO projectDao;
+
+    @Autowired
+    private IProjectSecurityChecker projectSecurityChecker;
+
+    @Resource(name = "projectconstants")
+    private Properties messages;
+
+    @Autowired
+    private IPassThroughProjectDocumentReader passThroughProjectDocumentReader;
+
+    @Override
+    @Transactional
+    public String addPassThroughProject(String userid, IPassThroughProject project) throws QuadrigaStorageException {
+        return projectDao.addPassThroughProject(userid, project);
+    }
+
+    @Override
+    @Transactional
+    public String getInternalProjectId(String externalProjectid, String userid)
+            throws QuadrigaStorageException, NoSuchRoleException {
+
+        List<PassThroughProjectDTO> projectDTOs = projectDao.getExternalProjects(externalProjectid);
+
+        for (PassThroughProjectDTO projectDTO : projectDTOs) {
+            boolean isOwner = projectSecurityChecker.isProjectOwner(userid, projectDTO.getProjectid());
+            if (isOwner) {
+                return projectDTO.getProjectid();
+            }
+            boolean isCollaborator = projectSecurityChecker.isCollaborator(userid,
+                    RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN, projectDTO.getProjectid());
+            if (isCollaborator) {
+                return projectDTO.getProjectid();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public String callQStore(String xml, IUser user) throws QStoreStorageException, DocumentParserException {
+        String networkName = Constants.VOGONWEB_NETWORK_NAME;
+
+        String userid = user.getUserName();
+        String networkId = null;
+        try {
+            Document document = passThroughProjectDocumentReader.getXMLParser(xml);
+
+            String projectId = passThroughProjectDocumentReader.getProjectID(document, userid);
+
+            String workspaceId = passThroughProjectDocumentReader.getWorsapceID(document, projectId, userid);
+
+            String annotatedText = passThroughProjectDocumentReader.getAnnotateData(xml);
+            String responseFromQStore = networkManager.storeNetworks(annotatedText);
+            networkId = networkManager.storeNetworkDetails(responseFromQStore, user, networkName, workspaceId,
+                    INetworkManager.NEWNETWORK, "", INetworkManager.VERSION_ZERO);
+
+        } catch (ParserConfigurationException | SAXException | IOException | NoSuchRoleException
+                | QuadrigaStorageException | JAXBException | QuadrigaAccessException e) {
+            throw new DocumentParserException(e);
+
+        }
+        return networkId;
+
+    }
+
+    @Override
+    @Transactional
+    public String createWorkspaceForExternalProject(String externalWorkspaceId, String externalWorkspaceName,
```

Comment:
there is a cycle here. documentreader uses this class, and this class uses documentreader. needs to be reworked.